### PR TITLE
Allow multiple trees as input to Filter

### PR DESCRIPTION
**Pull request by SparshithNR**
* Using fs-merger
    1. Allow filter developers to reduce mergeTree or funneling entire array of trees to one tree and then pass it to filter
    2. Allows input array of trees as input.
* input/output :
    1. Provides good ergonomics and avoids appending the this.outputPath
* Test:
    1. Added additional tests to make sure that Filter can take more than one tree

Example:
```js
class Awk extends Filter {
  constructor(inputNode, search, replace, options) {
    options = options || {};
    super(inputNode, {
      annotation: options.annotation
    });
    this.search = search;
    this.replace = replace;
    this.extensions = ['txt'];
    this.targetExtension = 'txt';
  }
  processString(content, relativePath) {
    return content.replace(this.search, this.replace);
  };
}
module.exports = new Awk(['fixture','fixture_2'], 'test', 'real');
```
Since we are now allowing users to pass two paths/nodes. We avoid a uncessary merge/funnel.
Before to achive the above result we had to below sinppet. 
```js
let fullTree = mergeTree(['fixture','fixture_2'];
module.exports = new Awk(fullTree, 'test', 'real');
```

### CODE REVIEW

File: lib/dependencies.js [hunk 1]
```diff
@@ -143,7 +147,12 @@ module.exports = class Dependencies {
     for (let i = 0; i < dependencies.length; i++) {
       let depPath = path.normalize(dependencies[i]);
       if (!path.isAbsolute(depPath)) {
-        depPath = path.resolve(this.rootDir, fileDir, depPath);
+        for (let index = 0; index < this.rootDirs.length; index++) {
+          let tempPath = path.resolve(this.rootDirs[index], fileDir, depPath);
+          if (fs.existsSync(tempPath)) {
+            depPath = tempPath;
```
Comment (chriseppstein) [inline lib/dependencies.js]: This code seems weird to me. I don't see where the mapped paths from the fsmerger are taken into account. Also, there was a definitive file that was already accessed -- I don't think we should go looking for it again at this point. Instead, I think I'd rather require that `filePath` is absolute now and just throw an error if passed a relative path.

File: index.js [hunk 2]
```diff
@@ -162,21 +167,21 @@ function invalidationsAsPatches(invalidated, currentTree, nextTree) {
 }
 
 Filter.prototype.build = function() {
-  // @ts-ignore
-  let srcDir = this.inputPaths[0];
```
[the remark this answers is not in the document]
Comment (SparshithNR) [reply index.js]: I think we should bump a major version for sure. 
But this won't be breaking change right away. As the inputPaths[0] will still have the src if they aren't using the feature of having an array of trees as an input to the persistent filter.
If someone is subclassing they can read our new feature and start using this.fsMergerFiles.readFileSync(relativePath).

File: index.js [hunk 3]
```diff
@@ -190,8 +195,9 @@ Filter.prototype.build = function() {
   let instrumentation = heimdall.start('derivePatches', DerivePatchesSchema);
 
   let walkStart = process.hrtime();
-  let entries = walkSync.entries(srcDir);
-  let nextTree = FSTree.fromEntries(entries);
+  this.fsMergerFiles = new FSMerger(this.inputNodes);
```
[the remark this answers is not in the document]
Comment (SparshithNR) [reply index.js]: I think there was something missing when I tried in a constructor, let me check again.

File: index.js [hunk 4]
```diff
@@ -374,13 +405,24 @@ Filter.prototype.canProcessFile =
...
-  if (this.inputPaths === undefined) {
+  if (this.fsMergerFiles === undefined && this.inputPaths === undefined) {
     return false;
   }
-
-  // @ts-ignore
-  let srcDir = this.inputPaths[0];
-  let path = srcDir + '/' + relativePath;
+  let path = '';
+  if (this.fsMergerFiles) {
```
[the remark this answers is not in the document]
Comment (SparshithNR) [reply index.js]: The reason the initialization of `this.fsMergerFiles` because `broccoliNode.outputPath` is available only in the build step, which is needed for fsMerger to work effectively.

File: index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (SparshithNR) [reply index.js]: Addressed in new commit.

File: lib/dependencies.js (hunk 1 above)
Comment (SparshithNR) [reply lib/dependencies.js]: After conversation with @chriseppstein offline, setDependencies now must accept the absolute path. This will be a breaking change. And we will bump the major version for this change.

File: index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (stefanpenner) [reply index.js]: Yes this will be part of a major bump.

File: index.js [hunk 5]
```diff
@@ -410,17 +422,18 @@ module.exports = class Filter extends Plugin {
     }
   }
 
-  async processFile(srcDir, destDir, relativePath, forceInvalidation, isChange, instrumentation, entry) {
+  async processFile(srcDir, relativePath, forceInvalidation, isChange, instrumentation, entry) {
```
Comment (stefanpenner) [inline index.js]: Changing this signature by dropping `destDir` is a breaking change, as `processFile` is public API for plugin subclasses. I don't think we want to do that, otherwise upgrades will be extra complex for little benefit.

File: index.js [hunk 6]
```diff
@@ -410,17 +422,18 @@ module.exports = class Filter extends Plugin {
...
     let filter = this;
     let inputEncoding = this.inputEncoding;
     let outputEncoding = this.outputEncoding;
 
     if (inputEncoding === undefined)  inputEncoding  = 'utf8';
     if (outputEncoding === undefined) outputEncoding = 'utf8';
 
-    let contents = fs.readFileSync(srcDir + '/' + relativePath, {
+    let contents = fs.readFileSync(entry.fullPath, {
```
Comment (stefanpenner) [inline index.js]: `this.input`?

File: index.js [hunk 7]
```diff
@@ -121,6 +128,7 @@ module.exports = class Filter extends Plugin {
       loggerName += ' > [' + annotation + ']';
     }
 
+    this.fsMerger = new FSMerger(inputTrees);
```
Comment (stefanpenner) [inline index.js]: by placing this on the `this` we are making `fsMerger` public, is that intentional?

Comment (stefanpenner) [inline index.js]: if this is a subclass of broccoli-plugin, and broccoli-plugin has `this.input` and `this.output` why do we need `FSMerger` here?

File: test/test.js [hunk 8]
```diff
@@ -1491,6 +1491,136 @@ describe('Filter', function() {
 
     });
   });
+
+  describe('Takes different types input nodes', function() {
+    class TestFilter extends Filter {
+      constructor(nodes) {
```
Comment (stefanpenner) [inline test/test.js]: this `constructor` is uneeded

File: test/test.js [hunk 9]
```diff
@@ -1491,6 +1491,136 @@ describe('Filter', function() {
...
+    it ('input node with prefix', async function() {
+      input.write({
+        'example': {
+          'map.js': 'let broccoli = 0;',
+        }
+      });
+
+      let FSMergerObjectWithFileDest = {
+        root: input.path(),
+        getDestinationPath:  function (relativePath) {
```
Comment (stefanpenner) [inline test/test.js]: concise function can be used `getDestinationPath(relativePath) { ...`

File: test/test.js [hunk 10]
```diff
@@ -1491,6 +1491,136 @@ describe('Filter', function() {
...
+      });
+
+      let FSMergerObjectWithPrefix = {
+        root: path.join(input.path(), 'docs'),
+        prefix: 'documents'
+      };
+
+      let FSMergerObjectWithFileDest = {
+        root: path.join(input.path(), 'example'),
+        getDestinationPath:  function (relativePath) {
```
Comment (stefanpenner) [inline test/test.js]: ^

File: index.js [hunk 11]
```diff
@@ -104,8 +104,15 @@ module.exports = class Filter extends Plugin {
     return !!result;
   }
 
-  constructor(inputTree, options) {
-    super([inputTree], {
+  constructor(inputTrees, options) {
+    // Parsed array of trees to have either broccoliNode or string which is expected by broccoli-plugin
```
Comment (stefanpenner) [inline index.js]: This processing of `inputTrees` is surprising, why is this needed?

File: index.js [hunk 12]
```diff
@@ -158,21 +166,17 @@ module.exports = class Filter extends Plugin {
...
 
   async build() {
-    // @ts-ignore
-    let srcDir = this.inputPaths[0];
-    // @ts-ignore
-    let destDir = this.outputPath;
 
     if (this.dependencyInvalidation && !this.dependencies) {
-      this.dependencies = this.processor.initialDependencies(srcDir);
+      this.dependencies = this.processor.initialDependencies(this.inputPaths);
```
Comment (stefanpenner) [inline index.js]: Passing `this.inputPaths` here breaks the encapsulation we hope to to achieve with `this.input` `this.output`, is this needed or can the Dependency code be made aware of `this.input`

File: index.js [hunk 13]
```diff
@@ -239,27 +243,35 @@ module.exports = class Filter extends Plugin {
...
+        outputPath = this.getDestFilePath(outputPath, entry) || outputPath;
         let forceInvalidation = invalidated.includes(relativePath);
 
         this._logger.debug('[operation:%s] %s', operation, relativePath);
 
         switch (operation) {
           case 'mkdir': {
             instrumentation.mkdir++;
-            return fs.mkdirSync(outputPath);
+            return this.output.mkdirSync(outputPath, { recursive: true });
```
Comment (stefanpenner) [inline index.js]: This should not be recursive, the efficient patch based operation are crafted in such a way where no extra IO is required. Making this recursive can introduce unwanted IO, and also hide bugs where the list of operations we are applying was incomplete.

File: index.js [hunk 14]
```diff
@@ -239,27 +243,35 @@ module.exports = class Filter extends Plugin {
...
           } case 'rmdir': {
             instrumentation.rmdir++;
-            return fs.rmdirSync(outputPath);
+            return this.output.rmdirSync(outputPath);
           } case 'unlink': {
             instrumentation.unlink++;
-            return fs.unlinkSync(outputPath);
+            return fs.unlinkSync( this.outputPath + '/' + outputPath);
```
Comment (stefanpenner) [inline index.js]: This is breaking the encapsulation intended by `input` and `output`

File: index.js [hunk 15]
```diff
@@ -292,27 +304,27 @@ module.exports = class Filter extends Plugin {
...
     try {
       let result;
-      let srcPath = srcDir + '/' + relativePath;
 
       if (this.canProcessFile(relativePath, entry)) {
         stats.processed++;
         if (this._outputLinks[outputPath] === true) {
           delete this._outputLinks[outputPath];
-          fs.unlinkSync(outputPath);
+          fs.unlinkSync(destDir);
```
Comment (stefanpenner) [inline index.js]: If we don't use `this.output` we also break encapsulation here.

File: index.js [hunk 16]
```diff
@@ -292,27 +304,27 @@ module.exports = class Filter extends Plugin {
...
-        result = await this.processAndCacheFile(srcDir, destDir, entry, forceInvalidation, isChange, stats);
+        result = await this.processAndCacheFile(relativePath, entry, forceInvalidation, isChange, stats);
       } else {
         stats.linked++;
         if (isChange) {
-          fs.unlinkSync(outputPath);
+          fs.unlinkSync(destDir);
         }
-        result = symlinkOrCopySync(srcPath, outputPath);
+        result = symlinkOrCopySync(entry.fullPath, destDir);
```
Comment (stefanpenner) [inline index.js]: this is breaking encapsulation.

File: index.js [hunk 17]
```diff
@@ -292,27 +304,27 @@ module.exports = class Filter extends Plugin {
...
-          fs.unlinkSync(outputPath);
+          fs.unlinkSync(destDir);
         }
-        result = await this.processAndCacheFile(srcDir, destDir, entry, forceInvalidation, isChange, stats);
+        result = await this.processAndCacheFile(relativePath, entry, forceInvalidation, isChange, stats);
       } else {
         stats.linked++;
         if (isChange) {
-          fs.unlinkSync(outputPath);
+          fs.unlinkSync(destDir);
```
Comment (stefanpenner) [inline index.js]: this is breaking encapsulation.

File: index.js [hunk 18]
```diff
@@ -362,15 +374,15 @@ module.exports = class Filter extends Plugin {
 
   isDirectory(relativePath, entry) {
     // @ts-ignore
-    if (this.inputPaths === undefined) {
+    if (this.fsMerger === undefined) {
```
Comment (stefanpenner) [inline index.js]: when is this `undefined` ?

File: package.json [hunk 19]
```diff
@@ -43,8 +43,8 @@
     "promise-map-series": "^0.2.1",
     "rimraf": "^3.0.0",
     "symlink-or-copy": "^1.0.1",
-    "sync-disk-cache": "^2.0.0",
-    "walk-sync": "^2.0.2"
+    "sync-disk-cache": "^1.3.3",
```
Comment (stefanpenner) [inline package.json]: why is this downgraded?

File: index.js (hunk 18 above)
Comment (SparshithNR) [reply index.js]: If this.isDirectory is called before `build` method is called.

File: index.js (hunk 17 above)
Comment (SparshithNR) [reply index.js]: We are yet to add the `unlinkSync` to `broccoli-plugin`. 
PRs: 
https://github.com/broccolijs/broccoli-plugin/pull/46
https://github.com/broccolijs/broccoli-plugin/pull/47

File: index.js (hunk 16 above)
Comment (SparshithNR) [reply index.js]: We are yet to add the `symlinkOrCopySync` to `broccoli-plugin`.

File: index.js (hunk 15 above)
Comment (SparshithNR) [reply index.js]: Yet to add `unlinkSync` to `this.output`

File: index.js (hunk 7 above)
Comment (SparshithNR) [reply index.js]: Because we want to support 
https://github.com/broccolijs/broccoli-persistent-filter/blob/9e01bdae7da04f6f7146c4dee1818d2fe75e1339/test/test.js#L1524 case.
`this.input` is a `broccoli-plugin` feature which expects `this.inputNodes` to be broccoli node or a string.

Comment (SparshithNR) [reply index.js]: I think we can make it private.

File: index.js (hunk 18 above)
Comment (SparshithNR) [reply index.js]: Removed it

File: index.js [hunk 20]
```diff
@@ -104,8 +101,19 @@ module.exports = class Filter extends Plugin {
     return !!result;
   }
 
-  constructor(inputTree, options) {
-    super([inputTree], {
+  get fsMerger() {
```
Comment (stefanpenner) [inline index.js]: as implemented `fsMerger` is still  part of the public API for subclasses, this seems sorta strange right?

File: index.js (hunk 7 above)
Comment (stefanpenner) [reply index.js]: Was that mistake, should this have been a primitive features of `this.input`? Or should the code in question operate on `this.input` rather then `inputTrees` directly?

Currently I am worried, that this plugin does not use `this.input`, did we make a mistake with it? By not using `this.input` we are making it hard for future optimizations, as ensuring all plugins go through `this.input` `this.output` opens many doors.

Comment (SparshithNR) [reply index.js]: Should we consider handling this case separately and use input/output facade changes first and then see what we can to extend that feature?

**Comment by SparshithNR** [comment]
Input output facade is addressed in #187.
We will address allowing multiple trees as input in this PR.

File: index.js (hunk 7 above)
Comment (stefanpenner) [reply index.js]: Unfortunately mixing input + the custom fsMerger object, will break encapsulation for subclasses of filter. Are they supposed to read from `this.input` or from the other?

Our goal here is, to ensure `this.input` is the usable for subclasses, as that is our new public API for reading from input trees.

File: index.js [hunk 21]
```diff
@@ -292,27 +306,29 @@ module.exports = class Filter extends Plugin {
...
-  async _handleFile(relativePath, srcDir, destDir, entry, outputPath, forceInvalidation, isChange, stats) {
+  async _handleFile(relativePath, entry, outputPath, forceInvalidation, isChange, stats) {
     stats.handleFile++;
 
     let handleFileStart = process.hrtime();
+    let destDir = this.outputPath + '/' + outputPath;
     try {
       let result;
-      let srcPath = srcDir + '/' + relativePath;
+      let srcDir = this.inputPaths[0]; // keeping this line to maintain the signature of the fn.
```
Comment (stefanpenner) [inline index.js]: Using `this.inputPaths[0]` here is breaking encapsulation, post `this.input` and `this.output` migration it should be possible to use only the new APIs, what features are missing or different approaches are required to make this statement true.

File: index.js (hunk 13 above)
Comment (stefanpenner) [reply index.js]: @SparshithNR  ^^

File: index.js (hunk 12 above)
Comment (stefanpenner) [reply index.js]: ^^

File: index.js [hunk 22]
```diff
@@ -121,6 +129,10 @@ module.exports = class Filter extends Plugin {
       loggerName += ' > [' + annotation + ']';
     }
 
+    FSMERGER.set(this, {
```
Comment (stefanpenner) [inline index.js]: Currently I am worried, that this plugin does not use `this.input`, did we make a mistake with it? By not using `this.input` we are making it hard for future optimizations, as ensuring all plugins go through `this.input` `this.output` opens many doors when it comes to future optimizations.

Comment (SparshithNR) [reply index.js]: We may have to make changes to the `fs-merger` library to somehow match inputPaths and input to `broccoli-persistent-filter` where we expect input can be `broccoli-node` `string` or an `object with prefix/getDestinationPath` function.

**Comment by SparshithNR** [comment]
Closing this as #188 is addressing this
